### PR TITLE
Preprobe info for known crate type

### PR DESCRIPTION
**Pull request by matklad**
Previously, we've calculated the set of crate types to learn about by
recursively walking the graph of units. However, to actually know
dependencies of a unit exactly, we must know target specific info, and
we don't know it at this moment (in fact, we are trying calculating it).

Note that crate-type calculation is already lazy, we don't have to calc
all crate-types upfront. So, let's just scrape this info once for
well-known crate types, and fill whatever is left lazily.

@alexcrichton would this approach work at all? I think it would, if `KNOWN_CRATE_TYPES` are all available for all target-tripples we support. Is it a valid assumption? 

The larger picture is that I am trying to make unit dependency resolution eager and move it into the separate file. I even got something working, but I have to run dependency resolution three times, because it is not exactly idempotent for various reasons, including this target-info stuff :) 

```
    cx.prepare()?;
    cx.build_unit_map(units.clone())?; // resolve dependencies
    cx.probe_target_info(&units)?;
    cx.build_unit_map(units.clone())?; // resolve again
    cx.build_used_in_plugin_map(&units)?;
    custom_build::build_map(&mut cx, &units)?;
    cx.build_unit_map(units.clone())?; // and resolve one final time :) 
```

### CODE REVIEW

**Comment by alexcrichton** [comment]
So historically Cargo strove to be compatible with multiple versions of rust and I think this sort of code was added around the time that we added the `cdylib` crate type. A lot of time has passed since then and we've pretty much solidified since then that Cargo only works with the `rustc` it's shipped with, no others. So in that sense, I think the original motivation for code like this (dynamically discovering crate types) has since gone away.

Now that being said I think a construction like this may make it tricky to add new crate types over time. For example if we add one on nightly Cargo would still be tested against stable/beta/nightly compilers and it may fail due to the unknown crate type on older compilers. With *that* being said it's not like we plan on adding crate types any time soon so I think it's fine to just deal with it somehow if it comes up.

Now with *all of that* being said I've long wanted to overhaul this. Running `rustc --print` is actually pretty expensive and is routinely the most expensive part of a null build from profiling I've done. In that sense I'd love to not run this at all and simply have a different way of figuring out this information like storing it internally or just looking for a bunch of candidates.

Anyway, with all that in mind, I'm totally ok landing this. It may cause trouble down the road but this is all pretty crazy code anyway. Does that sound ok to you @matklad?

**Comment by matklad** [comment]
So, the nice part here is that these KNOWN_CRATE_TYPES  can be only a subset of all crate types. That is, if we are to add a new crate type to rustc, it should work with Cargo out of the box, at the cost of additional rustc invocation if this new crate type is actually used. And when the new crate type is graduated to stable, we can put it in the list of known types. The only problem here is that this dynamic code path is very rare, and untested, but that’s probably OK.

I however am a bit skeptical about hard-coding this and avoiding rustc for crate types specifically. I think it’s the number of rustc invocations that matters, and when we learn about crate types, we learn a bunch of other info as well, and, unless we hard-code that info as well, hardcoding crate goes won’t actually help.

It would be interesting to look into caching rustc invocations maybe? Like, key the cache by rust —version, to reduce everything to a single rustc invocation, and ask for compiler version very early in the build and asynchronously, so as to mask latency.


TL;DR, I don’t think this code actually does make things harder down the road, because we still have the lazy code path.

**Comment by matklad** [comment]
Just realized that laziness I am talking about is not obvious from the pr itself :)


So, here’s the bit where we call rustc for crate types we haven’t hit during initial probing: https://github.com/matklad/cargo/blob/9f7618b36c07e91f10ce643fc6538a289f3c8935/src/cargo/ops/cargo_rustc/context.rs#L673

**Comment by alexcrichton** [comment]
Oh oops sorry about that, I forgot we had that! In that case I think we can definitely land this...

@bors: r+

Also I think it's a great idea to cache this info based on `rustc --version`. That'll handle `rustup` nicely I think
